### DISCUSSION TITLE
[GAME] Überarbeiten der Levelgraphen

### DISCUSSION THREAD

**Discussion by AMatutat** [Programmiermethoden/Dungeon#1390] [Ideas]
Wir verwenden Graphen, um unsere Level darzustellen. Dabei stellen die Knoten die Räume im Level dar, während die Kanten anzeigen, welche Räume durch Türen miteinander verbunden sind.

*Anmerkung:* Ein Knoten kann maximal 4 Kanten haben. Siehe auch: https://github.com/Programmiermethoden/Dungeon/issues/902

Bei der Generierung der Level steht die Erstellung des Levelgraphen an erster Stelle. Dazu wird für jeden gewünschten Raum (z.B. wenn das Level 10 Räume haben soll) ein Knoten erstellt. Anschließend wird der Wurzelknoten (Root-Knoten) festgelegt. Die Knoten werden dann miteinander verbunden, indem folgender Prozess durchgeführt wird:

1. Den Root-Knoten auswählen und eine zufällige Richtung (Norden, Osten, Süden, Westen) festlegen.
2. Prüfen, ob der Root-Knoten bereits eine Verbindung in diese Richtung hat:
   - Wenn nein, wird der Knoten mit dem Root-Knoten verbunden. Die Verbindung wird immer gegenüberliegend gezogen, d.h., Root Norden wird mit dem neuen Knoten Süden verbunden.
   - Wenn ja, den Kindsknoten als neuen Root-Knoten betrachten und die Schritte 1 und 2 wiederholen.
3. Wiederhole diese Schritte, bis alle Knoten verbunden sind.

Dieser Prozess ergibt einen Baum. Ursprünglich wurden zufällige Kanten in diesen Baum hinzugefügt, um aus dem Baum einen Graphen zu erstellen und sicherzustellen, dass die Level nicht zu viele Sackgassen enthalten. Allerdings waren die zufällig gezogenen Kanten problematisch, da sie die Level unübersichtlich und verwirrend gestalteten. Es konnte beispielsweise vorkommen, dass man von Raum A nach links in Raum B ging und dann von Raum B erneut nach links in Raum C ging, und von Raum C wiederum nach links zurück zu Raum A. Dadurch konnte man sich keine klare Karte des Levels im Kopf erstellen.

Es ist notwendig, Regeln für das Hinzufügen von Kanten festzulegen, um die Struktur der Level zu verbessern. Beispielsweise könnten Regeln wie "Eine Kante darf Knoten miteinander verbinden, die maximal eine Baum-Tiefe höher oder tiefer liegen" hilfreich sein.

Um die Orientierung zu vereinfachen und etwas mehr Abwechslung einzubringen, wäre es cool, wenn beim Generieren der richtigen Level von den Graphenverbunden dafür gesorgt wird, dass Knoten eines Graphen immer das gleiche Design haben (aktuell hat der gesamte Graphenverbund das gleiche Design), die einzelnen Graphen aber unterschiedliche Designs verwenden.

Und wenn wir schon dabei sind, sollte es immer einen Designunterschied beim Wechseln der Level-Teilbereiche geben. Also nach einem Ice-Level kommt kein anderes Ice-Level (direkt). Das ist eins zu eins das Graphen Coloring Problem.

**Comment by cagix** [comment]
@AMatutat das ding ist uralt. brauchen wir das noch?
